msgraph/core/middleware/retry.py

```diff
@@ -0,0 +1,223 @@
+
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.max_retries: int = min(
+            kwargs.pop('max_retries', self.DEFAULT_MAX_RETRIES), self.MAX_RETRIES
+        )
+        self.backoff_factor: float = kwargs.pop('retry_backoff_factor', self.DEFAULT_BACKOFF_FACTOR)
+        self.backoff_max: int = kwargs.pop('retry_backoff_max', self.MAXIMUM_BACKOFF)
+        self.timeout: int = kwargs.pop('retry_time_limit', self.MAX_DELAY)
+
+        status_codes: [int] = kwargs.pop('retry_on_status_codes', [])
+
+        self._retry_on_status_codes: set = set(status_codes) | self._DEFAULT_RETRY_STATUS_CODES
+        self._allowed_methods: set = frozenset(
+            ['HEAD', 'GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS']
+        )
+        self._respect_retry_after_header: bool = True
+
+    @classmethod
+    def disable_retries(cls):
+        """
+        Disable retries by setting retry_total to zero.
+        retry_total takes precedence over all other counts.
+        """
+        return cls(max_retries=0)
+
+    def get_retry_options(self, middleware_control):
+        """
+        Check if request specific configs have been passed and override any session defaults
+        Then configure retry settings into the form of a dict.
+        """
+        if middleware_control:
+            return {
+                'total':
+                min(middleware_control.get('max_retries', self.max_retries), self.MAX_RETRIES),
+                'backoff':
+                middleware_control.get('retry_backoff_factor', self.backoff_factor),
+                'max_backoff':
+                middleware_control.get('retry_backoff_max', self.backoff_max),
+                'timeout':
+                middleware_control.get('retry_time_limit', self.timeout),
+                'retry_codes':
+                set(middleware_control.get('retry_on_status_codes', self._retry_on_status_codes))
+                | set(self._DEFAULT_RETRY_STATUS_CODES),
+                'methods':
+                self._allowed_methods,
+            }
+        return {
+            'total': self.max_retries,
+            'backoff': self.backoff_factor,
+            'max_backoff': self.backoff_max,
+            'timeout': self.timeout,
+            'retry_codes': self._retry_on_status_codes,
+            'methods': self._allowed_methods,
+        }
+
+    def send(self, request, **kwargs):
+        """
+        Sends the http request object to the next middleware or retries the request if necessary.
+        """
+        retry_options = self.get_retry_options(request.context.middleware_control)
+        absolute_time_limit = min(retry_options['timeout'], self.MAX_DELAY)
+        response = None
+        retry_count = 0
+        retry_valid = True
+
+        while retry_valid:
+            try:
+                start_time = time.time()
+                request.headers.update({'retry-attempt': '{}'.format(retry_count)})
```

Comment:
We are sending the retry-attempt header before we've retried. Let's only send this when we retry_count > 0. That way the existence of retry-attempt can only mean that a retry scenario has occurred.

Reply:
This makes sense. I'll update accordingly.